oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenIntrospectionAuthenticationProvider.java

```diff
@@ -0,0 +1,113 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.server.authorization.authentication;
+
+import static org.springframework.security.oauth2.server.authorization.authentication.OAuth2AuthenticationProviderUtils.getAuthenticatedClientElseThrowInvalidClient;
+
+import org.springframework.security.authentication.AuthenticationProvider;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.oauth2.core.AbstractOAuth2Token;
+import org.springframework.security.oauth2.jwt.JwtClaimNames;
+import org.springframework.security.oauth2.server.authorization.OAuth2Authorization;
+import org.springframework.security.oauth2.server.authorization.OAuth2Authorization.Token;
+import org.springframework.security.oauth2.server.authorization.OAuth2AuthorizationService;
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClient;
+import org.springframework.util.Assert;
+
+import java.time.Instant;
+import java.util.Map;
+
+/**
+ * An {@link AuthenticationProvider} implementation for OAuth 2.0 Token Introspection.
+ *
+ * @author Gerardo Roza
+ * @since 0.1.1
+ * @see OAuth2TokenIntrospectionAuthenticationToken
+ * @see OAuth2AuthorizationService
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7662#section-2.1">Section 2.1 - Introspection Request</a>
+ */
+public class OAuth2TokenIntrospectionAuthenticationProvider implements AuthenticationProvider {
+	private final OAuth2AuthorizationService authorizationService;
+
+	/**
+	 * Constructs an {@code OAuth2TokenIntrospectionAuthenticationProvider} using the provided parameters.
+	 *
+	 * @param authorizationService the authorization service
+	 */
+	public OAuth2TokenIntrospectionAuthenticationProvider(OAuth2AuthorizationService authorizationService) {
+		Assert.notNull(authorizationService, "authorizationService cannot be null");
+		this.authorizationService = authorizationService;
+	}
+
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		OAuth2TokenIntrospectionAuthenticationToken tokenIntrospectionAuthentication = (OAuth2TokenIntrospectionAuthenticationToken) authentication;
+
+		OAuth2ClientAuthenticationToken clientPrincipal = getAuthenticatedClientElseThrowInvalidClient(
+				tokenIntrospectionAuthentication);
+		RegisteredClient registeredClient = clientPrincipal.getRegisteredClient();
+
+		OAuth2Authorization authorization = this.authorizationService
+				.findByToken(tokenIntrospectionAuthentication.getTokenValue(), null);
+		if (authorization == null) {
+			return generateAuthenticationTokenForInvalidToken(clientPrincipal, registeredClient);
+		}
+
+		if (!registeredClient.getId().equals(authorization.getRegisteredClientId())) {
```

Comment:
This check should be removed. Any authenticated client could introspect a token that was not originally granted to it.

Reply:
I now removed the check @jgrandja but keep in mind that even though the RFC spec doesn't seem to suggest any approach regarding this, Okta seems to have a clear position towards this, which is not allowing clients to introspect others' tokens:
https://www.oauth.com/oauth2-servers/token-introspection-endpoint/

> Any other error is considered an “inactive” token.
>*  The token requested does not exist or is invalid
>*  The token expired
>*  **The token was issued to a different client than is making this request**
>
>In any of these cases, it is not considered an error response, and the endpoint returns simply an inactive flag.

Reply:
Thanks for being on top of this @rozagerardo ! Can you check to see if Keycloak, Curity and Microsoft Azure enforce this policy? Then we can decide what to do.

Right now I'm thinking we should re-apply the same policy. It's always best to start off more secure than less. We can always open things up later if it makes sense.

Reply:
No problem @jgrandja , IMO it's part of the commitment and the ownership I assume when I request working on a feature 👍 

BTW, sorry for the delay with this, I didn't have much free time these last weeks.

Ok, regarding my findings, it seems **all** providers (surprisingly, even Okta) do allow to introspect tokens issued to a different client.

Here are the demo videos for the three providers (Keycloak, Okta and Curity, as Azure AD doesn't yet provide support for the Introspection Endpoint - as indicated [here](https://feedback.azure.com/forums/169401-azure-active-directory/suggestions/18930289-introspection-endpoint-for-azure-active-directory)) showing how they enforce this policy, just in case you want to see this in action:
https://drive.google.com/drive/folders/1LC98BnxTWDl82hCp3RYBWq9KoZT5n2nz?usp=sharing

Also, just FYI I reported a question in the Okta support center just to clarify if they should update the documentation:
https://devforum.okta.com/t/introspection-endpoint-should-work-for-tokens-issued-by-different-client/13587

Reply:
Section 4. Security Considerations of [RFC 7662] OAuth 2.0 Token Introspection discusses the token validation steps required by the AS when processing the Introspection request:

```
   o  If the token can expire, the authorization server MUST determine
      whether or not the token has expired.
   o  If the token can be issued before it is able to be used, the
      authorization server MUST determine whether or not a token's valid
      period has started yet.
   o  If the token can be revoked after it was issued, the authorization
      server MUST determine whether or not such a revocation has taken
      place.
   o  If the token has been signed, the authorization server MUST
      validate the signature.
   o  If the token can be used only at certain resource servers, the
      authorization server MUST determine whether or not the token can
      be used at the resource server making the introspection call.
      
   If an authorization server fails to perform any applicable check, the
   resource server could make an erroneous security decision based on
   that response.  Note that not all of these checks will be applicable
   to all OAuth 2.0 deployments and it is up to the authorization server
   to determine which of these checks (and any other checks) apply.
   
   If left unprotected and un-throttled, the introspection endpoint
   could present a means for an attacker to poll a series of possible
   token values, fishing for a valid token.
```

Based on the last of the AS Token Endpoint validation requirements, IMO you've discovered potential security issues with Keycloak, Okta, and Curity AS software.

Reply:
Thanks for the comment @dfcoffin.

I interpreted the last validation requirement differently, IMO the _client_id_ doesn't relate strictly with the _resource servers_ in the sense of serving to "identify" a particular Resource Server (in fact, it would be used to identify a particular Client, but not the Resource Server, which is the one consuming this endpoint. Read the last section about Curity to know how they modeled this). In fact, the RFC gives a sort of example of how this Resource Server whitelist/blacklist would look like:

> The introspection endpoint MAY accept other OPTIONAL parameters to
   provide further context to the query.  For instance, an authorization
   server may desire to know the IP address of the client accessing the
   protected resource to determine if the correct client is likely to be
   presenting the token.

In my opinion, this is what the restriction refers to, a sort of IP-based firewall to restrict requests from just any origin.

Of course, I might be getting this wrong, let me know what you think.

Now, just to add some more information, in the case of Curity, it is definitely a conscious decision, on their platform you have to enable explicitly the introspection functionality granularly on an application/client, and they suggest not mixing this feature with other token-issuer clients; they separate this functionality as it is (commonly) used by Resource Servers instead of Clients:


> https://curity.io/resources/learn/introspect-with-phantom-token/#configure-client-for-introspection-gateway-client

Reply:
Awesome, thanks for the very interesting feedback @dfcoffin :)

Yes, the Client names used in the videos might not be totally representative, as my point was only to analyze if a Client (of any type) could introspect tokens issued by another Client (again, of any type, or with any enabled feature).

Anyway, IMO the feedback provided by Justin strengthens the idea that is ok to authenticate using a particular Client's credentials to introspect another Client's tokens, as a Client setup that is configured to issue tokens is not necessarily the same that will introspect the token.

This is especially clear for Public Clients; they would need a different, confidential Client to be able to Introspect its tokens.

Reply:
@dfcoffin @rozagerardo Regarding the following reference in RFC 7662:

> If the token can be used only at certain resource servers, the
   authorization server MUST determine whether or not the token can
   be used at the resource server making the introspection call.

This seems very related to how [Resource Indicators for OAuth 2.0](https://tools.ietf.org/html/rfc8707) would be implemented.

As far as Curity goes:

> Introspection clients are rarely the same as clients that can request tokens. A gateway client should be setup to only be allowed to do introspection and no other OAuth flow.

This makes total sense to me and I really do like how Curity has implemented their product.

The 2 configuration options that I would favour for an introspection-specific client is Mutual TLS client authentication and Bearer Token client authentication. The bearer token would have introspection-specific scopes associated with it (e.g. `token.introspect`) and a client certificate claim (e.g. `x5t#S256` - X.509 Certificate SHA-256 Thumbprint).

Reply:
Thanks for the comments @jgrandja!

Let's see, regarding the first note about Resource Indicators for OAuth 2.0, that is indeed very interesting, and it seems it would be related, but I'm having some doubts.

I just had a quick look at RFC 8707 so I might be missing a few things, but basically, it seems this mechanism would impact mainly in the "audience" value the Authorization Server retrieves:

> The authorization server **SHOULD audience-restrict** issued access
   tokens to the resource(s) indicated by the "resource" parameter.
   Audience restrictions can be communicated in JSON Web Tokens
   [RFC7519] **with the "aud" claim and the top-level member of the same
   name provides the audience restriction information in a Token
   Introspection** [RFC7662] response.
   
So, going back to the original note in the Token Introspection spec:

>  o  If the token can be used only at certain resource servers, the
      authorization server MUST determine whether or not the token can
      be used at the resource server making the introspection call.
	  
I'm not sure a check based on the RFC 8707 feature corresponds to the one mentioned above, if the first impacts the audience, then IMO it should be the Resource Server the one deciding if the token represents authorized access to its resource...We could consider the following behavior though: 

- if the _resource_ is a "_https URL_" then it might be suitable to perform the validation right away in the AS Introspection Endpoint and consider the token invalid if introspected by a different Resource Server
- if the _resource_ is a "_more abstract identifier_", then simply use it as the "aud" field and consider the token "active" 

But this would be whole new discussion and we shouldn't loose focus :)

Finally, regarding the two configuration options you mention, yes, I think that makes sense; an introspection-specific client would mandatorily be a confidential client thus it would be inherently suitable for these authentication methods... I don't totally understand the use of forcing the client to add a particular scope in the Bearer token for this, but I haven't used this mechanism much, so I'd love to see/participate in the development of that feature.

Now, regarding the next steps here, we don't yet have support for these authentication methods, right? So, maybe we need to create a few tickets (maybe to allow setting introspection as a feature on client configurations, analyze implementing Resource Indicators, and to support different methods of client authentication), but is there anything that should be part of this particular ticket?

Looking forward to your comments regarding this Joe (and @dfcoffin, of course), and let me know if you want me to create tickets for some of the features discussed here 👍 

PS: I found very appealing the solution used by Curity too :)

Reply:
@jgrandja @rozagerardo 

> Now, regarding the next steps here, we don't yet have support for these authentication methods, right? So, maybe we need to create a few tickets (maybe to allow setting introspection as a feature on client configurations, analyze implementing Resource Indicators, and to support different methods of client authentication), but is there anything that should be part of this particular ticket?

RFC 8707 only references "aud" once and JWT three times in a single paragraph. On the other hand, it references RFC 6749 twelve times.  Based on the occurrences of RFC6749 and the inclusion in the introduction of the following:

> Bearer tokens, currently the most commonly utilized type of OAuth access token, allow any party in possession of a token to get access to the associated resources.  To prevent misuse, several important security assumptions must hold, one of which is that an access token must only be valid for use at a specific protected resource and for a specific scope of access.  Section 5.2 of [RFC6750], for example, prescribes including the token's intended recipients within the token to prevent token redirect.  When the authorization server is informed of the resource that will process the access token, it can restrict the intended audience of that token to the given resource such that the token cannot be used successfully
> at other resources.

IMO we need to list RFC 8707 as an EPIC since its implementation will require modification of several existing filters and endpoints to determine if the "resource" parameter is present in the Authorization Request and how to handle it.  Adding a "TODO" flag to review when RFC 8707 Resouce Indicators for OAuth 2.0 is implemented.

Reply:
Right, that was kind of where I was heading too, IMO these topics have an impact that goes beyond this current feature, and I don't think anything in the code involved here should be changed (at this point).

Let me know if you want me to create a ticket for any of the issues discussed here, and just to be clear regarding this @dfcoffin:

> Adding a "TODO" flag to review when RFC 8707 Resouce Indicators for OAuth 2.0 is implemented.

I'm not totally sure what you mean, is that referring just to a personal reminder you have? I mean, that is not something we should add to the code right? (I've seen a just couple of TODO flags, but I think this shouldn't be the case). Sorry about this doubt, I'm not totally familiarized with the tickets/features management process you guys follow :)

cc @jgrandja 

PS Donald, note that the section you quoted about Bearer tokens also mentions that the _Resource Indicators_ feature is carried out by restricting the audience of the token :)

Reply:
@rozagerardo @dfcoffin Thanks for all your feedback.

I just logged a ticket for Resource Indicators gh-279.

To recap on how I view Resource Indicators will be implemented:

1) The client sends the `resource` parameter in the authorization or token request. Its value can be 1 or more absolute URI's.
2) The authorization server will use the `aud` claim to represent the value(s) in the `resource` parameter. The authorization server MAY also apply some policy decisions around the resource and further enhance the token.
3) The resource server will use the `aud` claim to apply authorization decisioning on whether to allow the request to proceed. This is where it checks if it's the "intended audience" of the incoming token.

@rozagerardo Resource indicators is a separate feature so there is no additional work for this PR. Apologies for the delay on this PR but I've been quite busy on the OAuth Client side of things as we're releasing Spring Security 5.5.0-RC1 today. My time has now opened up and I'm refocusing back on Authorization Server starting today. I'm circling back to this PR with the plan on merging it before end of week.

Regarding client authentication method support, this will be a major theme for `0.1.2`. See gh-59 and gh-101

Reply:
Cool, everything sounds good to me @jgrandja thanks for the update!

Don't worry for the times BTW, I know you have a lot on your plate and I also understand that new projects are not totally stable in the early stages; keeping the PR updated does require some extra effort, but that's totally expected and is part of my commitment towards the project :)

Of course, let me know if I can help with anything else, either with this ticket or with any other task; I'll probably pick up a new ticket at some point, but some pressure to grab a new task works for me too :) cheers!